### PR TITLE
Implemented iqs5xx driver

### PR DESCRIPTION
**Pull request by o7-machinehum**
This PR is to implement the driver and API for the iqs5xx trackpad controller

### CODE REVIEW

**Comment by carlescufi** [comment]
@Machine-Hum thanks for your contribution. Please format your commits according to our guidelines:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow

Remember to amend existing commits, not add new ones. Also you will need to remove the merge commit you have now.

**Comment by Machine-Hum** [comment]
Should those twister/clang CI tests be passing? I don't see where I would have broken anything.

If there is something else I should be doing or reading please let me know :)

**Comment by carlescufi** [comment]
> @carlescufi I think things should look a little better now. Formatted as per the guidelines and ran the clang linter. I tried running twister and it took over 24hrs to get to 75%. Is this normal?

You tried to run twister locally? Yes, that's gonna take a long time. Use CI instead of running locally.

File: drivers/touch/Kconfig [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2022 Interrupt Labs <info@interruptlabs.ca>
+# SPDX-License-Identifier: Apache-2.0
+menuconfig TOUCH
+	bool "Touch Drivers"
+	help
+		Include touch drivers in system config
```
Comment (talih0) [inline drivers/touch/Kconfig:6]: the indent style is <Tab><space><space> for this line, same as you have below.

File: drivers/touch/Kconfig [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	int "Touch init priority"
+	default 90
+	help
+	  Touch initialization priority.
+
+config TOUCH_DEBUG
+	bool "Touch Debugging"
+	default n
+	help
+		This will print out touch events and other debugging info.
```
Comment (talih0) [inline drivers/touch/Kconfig:19]: indent style

File: drivers/touch/iqs5xx/Kconfig [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+config TOUCH_LOG_LEVEL
+	int "Log Level for Touch"
+	default 4
+
+config IQS5XX
+	bool "IQS5XX Driver"
+	depends on I2C && GPIO
+	help
+		Projected capacitive trackpad/touchscreen controller.
```
Comment (talih0) [inline drivers/touch/iqs5xx/Kconfig:14]: indent style

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	return ret;
+}
+
+static int iqs5xx_read_word(const struct i2c_dt_spec *client, uint16_t reg, uint16_t *val)
+{
+	uint16_t val_buf;
+	int error;
+
+	error = iqs5xx_read_burst(client, reg, &val_buf, sizeof(val_buf));
+	if (error)
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:51]: zephyr suggests braces around each if, else, for, etc, statement

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+static int iqs5xx_write_burst(const struct i2c_dt_spec *client, uint16_t reg, const void *val,
+				  uint16_t len)
+{
+	int ret, i;
+	uint16_t mlen = sizeof(reg) + len;
+	uint8_t mbuf[sizeof(reg) + IQS5XX_WR_BYTES_MAX];
+
+	if (len > IQS5XX_WR_BYTES_MAX)
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:66]: braces around if

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+	 */
+	for (i = 0; i < IQS5XX_NUM_RETRIES; i++) {
+		ret = i2c_write_dt(client, mbuf, mlen);
+		if (ret >= 0)
+			return 0;
+
+		k_sleep(K_USEC(200));
+	}
+
+	if (ret >= 0)
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:85]: probably you meant ret < 0? 
also missing braces around if and line 79

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+	const struct iqs5xx_dev_config *cfg = data->dev->config;
+
+	iqs5xx_read_burst(&cfg->i2c, IQS5XX_GEST_EV0, &data->regmap, sizeof(data->regmap));
+
+	iqs5xx_write_byte(&cfg->i2c, IQS5XX_END_COMM, 0);
+
+	/* Small delay required as the iqs twiddles the rdy pin. */
+	k_sleep(K_USEC(50));
+
+#ifdef CONFIG_TOUCH_DEBUG
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:134]: maybe remove ifdef and use LOG_DGB in logging instead?

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+		case IQS5XX_ZOOM:
+			LOG_INF("Zoom");
+			break;
+		default:
+			LOG_INF("%d", data->regmap.gesture_event[1]);
+		}
+	}
+#endif
+
+	data->input_serviced = false;
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:177]: looks that input_serviced is not used anywhere else

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+	struct iqs5xx_data *data = dev->data;
+	const struct iqs5xx_dev_config *cfg = dev->config;
+	int rc;
+
+	if (!device_is_ready(cfg->rdy_gpio.port)) {
+		LOG_ERR("%s: device %s is not ready", dev->name, cfg->rdy_gpio.port->name);
+		return -ENODEV;
+	}
+
+	rc = gpio_pin_configure_dt(&cfg->rdy_gpio, GPIO_INPUT | cfg->rdy_gpio.dt_flags);
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:192]: nit: you've used variable names ret, rc, error in different places. Maybe just use ret everywhere?

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+static int iqs5xx_two_finger_tap(const struct device *dev)
+{
+	struct iqs5xx_data *data = dev->data;
+
+	if (data->regmap.gesture_event[0] & IQS5XX_TWO_FINGER_TAP) {
+		data->regmap.gesture_event[0] &= ~IQS5XX_TWO_FINGER_TAP;
+		return true;
+	}
+
+	return false;
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:219]: maybe the function return should be bool type then? same for the other API functions

File: drivers/touch/iqs5xx/iqs5xx.h [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	struct gpio_callback gpio_cb;
+	const struct device *dev;
+	bool input_serviced;
+	struct iqs5xx_regmap regmap;
+	struct k_work work;
+};
+
+struct iqs5xx_dev_config {
+	struct i2c_dt_spec i2c;
+	struct gpio_dt_spec rst_gpio;
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.h:126]: rst_gpio is unused

File: drivers/touch/iqs5xx/iqs5xx.c [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+static const struct touch_driver_api iqs5xx_driver_api = {
+	.single_tap = &iqs5xx_single_tap,
+	.two_finger_tap = &iqs5xx_two_finger_tap,
+	.x_pos_abs = &iqs5xx_x_position_abs,
+	.y_pos_abs = &iqs5xx_y_position_abs,
+	.x_pos_rel = &iqs5xx_x_position_rel,
+	.y_pos_rel = &iqs5xx_y_position_rel,
+	.num_fingers = &iqs5xx_num_fingers,
```
Comment (talih0) [inline drivers/touch/iqs5xx/iqs5xx.c:276]: Would it be useful to also have some callbacks to user when specific events occur?

**Review by gmarull — CHANGES_REQUESTED** [review]
I think we really need an input subsystem, not another API to handle a type of input device (touch).

**Comment by Machine-Hum** [comment]
Could someone rerun the workflows please :)

**Comment by o7-machinehum** [comment]
> I think we really need an input subsystem, not another API to handle a type of input device (touch).

So what's the action on my part? Should this be rewritten? I moved this to #52649 because I broke this PR, it's not updated from my remote. Closing.

**Pull request closed without merging**